substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java

```diff
@@ -78,24 +90,42 @@ public static UnsignedWord size() {
             throw VMError.shouldNotReachHere("Accessing the physical memory size may require allocation and synchronization");
         }
 
-        if (!isInitialized()) {
-            /*
-             * Multiple threads can race to initialize the cache. This is OK because all of them
-             * will (most-likely) compute the same value.
-             */
-            INITIALIZING.incrementAndGet();
-            try {
-                long memoryLimit = SubstrateOptions.MaxRAM.getValue();
-                if (memoryLimit > 0) {
-                    cachedSize = WordFactory.unsigned(memoryLimit);
-                } else {
-                    memoryLimit = Containers.memoryLimitInBytes();
-                    cachedSize = memoryLimit > 0
-                                    ? WordFactory.unsigned(memoryLimit)
-                                    : ImageSingletons.lookup(PhysicalMemorySupport.class).size();
+        synchronized (INITIALIZING) {
+            if (!isInitialized()) {
+                VMError.guarantee(INITIALIZING.get() <= 1, "Must not initialize twice");
+                if (isInitializing()) {
+                    /*
+                     * Recursive initializations need to wait for the one initializing thread to
+                     * finish so as to get correct reads of the cachedSize value.
+                     */
+                    try {
+                        boolean expired = !CACHED_SIZE_AVAIL_LATCH.await(1L, TimeUnit.SECONDS);
+                        if (expired) {
+                            throw new InternalError("expired latch!");
+                        }
+                        VMError.guarantee(cachedSize != UNSET_SENTINEL, "Expected chached size to be set");
+                        return cachedSize;
+                    } catch (InterruptedException e) {
+                        throw VMError.shouldNotReachHere("Interrupt on countdown latch!");
+                    }
+                }
+                INITIALIZING.incrementAndGet();
+                try {
+                    long memoryLimit = SubstrateOptions.MaxRAM.getValue();
+                    if (memoryLimit > 0) {
+                        cachedSize = WordFactory.unsigned(memoryLimit);
+                    } else {
+                        memoryLimit = Containers.memoryLimitInBytes();
+                        cachedSize = memoryLimit > 0
+                                        ? WordFactory.unsigned(memoryLimit)
+                                        : ImageSingletons.lookup(PhysicalMemorySupport.class).size();
+                    }
+                    // Now that we have set the cachedSize let other threads know it's
+                    // available to use.
+                    CACHED_SIZE_AVAIL_LATCH.countDown();
+                } finally {
+                    INITIALIZING.incrementAndGet();
```

Comment:
My understanding is that this will run even if line 118 throws an exception and `cachedSize` won't get set. 

Shouldn't we just fail initialization and return the state to uninitialized so that other parts can handle the issue? E.g. https://github.com/oracle/graal/blob/544b6f80d090bc320b69318efb325075de87e2a9/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java#L140

Alternatively shouldn't we set a fallback value for `cachedSize` and count down the latch in this case?

Reply:
This should be fixed in the updated version.